lib/mkCargoDerivation.nix

```diff
@@ -39,13 +40,55 @@ let
     "pnameSuffix"
     "stdenv"
   ];
+
+  rustcWrapper = writeShellApplication {
+    name = "rustc-wrapper";
+    text = ''
+      set -euo pipefail
+
+      args=("$@")
+
+      temp_dir=$(mktemp -d)
+      trap 'rm -rf -- "$temp_dir"' EXIT
+
+      for i in "''${!args[@]}"; do
+        if [[ "''${args[i]}" == --out-dir=* ]]; then
+          current_out_dir="''${args[i]#--out-dir=}"
+          args[i]="--out-dir=$temp_dir"
+          break
+        elif [[ "''${args[i]}" == --out-dir ]]; then
+          current_out_dir="''${args[i+1]}"
+          args[i+1]="$temp_dir"
+          break
+        fi
+      done
+
+      if [[ -v current_out_dir ]]; then
+        stderr="$temp_dir/crane_stderr"
```

Comment:
Why bother with capturing stderr and stdout?

Reply:
Oh. Cargo uses it to know when a process finished, instead of waiting for it to fully terminate?

:/

Reply:
Thought: `rm -Rf out-dir` before starting original, would take care of this, get rid of an EXIT hook, temp dir altogether. If it worked.

But it just occurred to me that it's `out-DIR` and probably oftentimes the same dir is used to write many files, so that probably doesn't fly.

Reply:
One could use `LD_PRELOAD` to intercept `open`, `write` syscalls in a rustc-wrapper and basically always follow the link if file is being read, and delete the symlink if being written to.

https://osterlund.xyz/posts/2018-03-12-interceptiong-functions-c.html

A bit involved, but could probably work perfectly and be 100% invisible.

Reply:
There's no good ld_preload equivalent on macos which is why i didn't go down that path - and that seems extremely brittle to recompiles. Ideally Nix would let us mount fuse filesystems inside a build - that'd be the 'reliable' solution to this.

I felt this traded off nicely between being robust to change but still getting the job done.

Reply:
Yeah, I don't see a better way.